### DISCUSSION TITLE
Accessing multiple entities in `QueryData`/`WorldQuery` impl

### DISCUSSION THREAD

**Discussion by BigWingBeat** [bevyengine/bevy#13971] [Q&A] [answered]
I have two component types, `Foo` and `Bar`. `Foo` holds an `Entity` value which always points to an entity with a `Bar` component, and this is a many-to-one relationship, so multiple `Foo` entities can exist that all point to the same `Bar` entity. The catch is that in order to do anything useful with a `Foo`, you also need a `&mut` reference to its corresponding `Bar`. As this is for a library, I want to hide this relationship as an implementation detail, so that users can simply query for `Foo`, and the library handles retrieving the corresponding `Bar` internally.

There are three ways I can think of to achieve this:
1) Don't bother, just push all the bookkeeping onto the user. Terrible UX.
2) Create a custom `SystemParam` type. Would require a lot of boilerplate manually duplicating Bevy's `Query` API.
3) Create a custom `QueryData`/`WorldQuery` type. I would prefer this, but I don't understand the ECS internals enough to determine if would be sound for the impl to fetch both the `Foo` value for the requested entity, as well as the `Bar` value from whatever entity is specified by the `Foo`.

Is it possible to soundly do 3, or is there a better solution?

**Comment by BigWingBeat** [comment]
(Cross-posted from Discord: https://discord.com/channels/691052431525675048/1253399316773736520)

**Comment by BigWingBeat** [comment, marked as answer]
Answer from Giuschi#7708 on Discord:

> I don't think this is possible to do with a query. `Query` assumes that you can get all its elements at once (e.g. with `query.iter_mut().collect()`). However if in your example you would yield a `&mut Foo` multiple times (one for each `Bar` connected to it), which is not sound according to Rust's rules.
You'll have to separately query for `&mut Foo`s and borrow from that query when you want to do something with a `Bar`. You could possibly wrap this in a custom `SystemParam` to make it a bit more ergonomic, but fundamentally you'll have to do the query in separate steps.
I don't think relationships will help because you likely won't be able to modify the relationship while iterating (only through commands for example), meaning no `&mut Foo` access.

TL;DR No it isn't sound, and 2 is the most reasonable solution.
